File: 05 May/05-07/longestCommonSubstring.js
function getLongestSubstring(str) {
  let longestSubstring = "";

  for (let substringLength = 1; substringLength < str.length; ++substringLength) {

    for (let index1 = 0; index1 < str.length - substringLength; ++index1) {
      const substring1 = str.slice(index1, index1 + substringLength);

      for (let index2 = index1 + 1; index2 <= str.length - substringLength; ++index2) {
        const substring2 = str.slice(index2, index2 + substringLength);

        if (substring1 === substring2 && substring1.length > longestSubstring.length) {
          longestSubstring = substring1;
          break;
        }
      }
    }
  }

  return longestSubstring;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. getLongestSubstring("abracadabra") should return "abra".
2. getLongestSubstring("hello world hello") should return "hello".
3. getLongestSubstring("mississippi") should return "issi".
4. getLongestSubstring("ha ha ha ha ha ha ha") should return "ha ha ha ha ha ha".
5. getLongestSubstring("the quick brown fox jumped over the lazy dog that the quick brown fox jumped over") should return "the quick brown fox jumped over".
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);

File: 05 May/05-07/longest_common_substring.py
def get_longest_substring(s):
    longest_substring = ""

    for substring_length in range(1, len(s)):

        for index1 in range(len(s) - substring_length):
            substring1 = s[index1 : index1 + substring_length]

            for index2 in range(index1 + 1, len(s) - substring_length + 1):
                substring2 = s[index2 : index2 + substring_length]

                if substring1 == substring2 and len(substring1) > len(longest_substring):
                    longest_substring = substring1
                    break

    return longest_substring

# --- TEST SUITE ---

tests_text = r'''
1. get_longest_substring("abracadabra") should return "abra".
2. get_longest_substring("hello world hello") should return "hello".
3. get_longest_substring("mississippi") should return "issi".
4. get_longest_substring("ha ha ha ha ha ha ha") should return "ha ha ha ha ha ha".
5. get_longest_substring("the quick brown fox jumped over the lazy dog that the quick brown fox jumped over") should return "the quick brown fox jumped over".
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)